### DISCUSSION TITLE
Support JSON with useFetcher.submit()

### DISCUSSION THREAD

**Discussion by byudaniel** [remix-run/remix#1959] [Proposals]
It would be handy to have the option to submit arbitrary JSON with `useFetcher.submit()`. I've got a very complex form with nested arrays that I could technically force everything into hidden fields and use `<Form ...>`, but for this use case feel that submitting JSON would be cleaner. Currently the behavior serializes an object into FormData, which is turning an array into `[object Object]`. 

Are there any plans for something like `useFetcher.submit({ myComplexData: [{...}] }, { method: 'post', encType: 'application/json' })`?

**Comment by ccssmnn** [comment]
Have you considered using the [JavaScript Fetch API](https://developer.mozilla.org/de/docs/Web/API/Fetch_API)?

**Reply by byudaniel** [reply]
Yes, I'm doing that as a workaround, but it would make sense to include it within the `useFetcher` API too.

I had come across [this thread](https://twitter.com/mjackson/status/1485299298128392192) and had assumed it was a supported scenario.

**Reply by ccssmnn** [reply]
From the Remix docs [here](https://remix.run/docs/en/v1/api/remix#fetchersubmit) you should be able to do this already. The advantage of this over the JavaScript Fetch API is, that you get all the utilities for building UI depending on your request.

```javascript
function SomeComponent() {
  const fetcher = useFetcher();

  const onClick = () =>
    fetcher.submit({ some: "values" }, { method: "post" });

  // ...
}
```

Is this what you were looking for? Otherwise you might need to specify your question.

**Reply by byudaniel** [reply]
`fetcher.submit({ some: "values" }, { method: "post" });` will serialize shallowly as FormData, so if you have a complex object it won't be serialized correctly. 

`useFetcher.submit({ myComplexData: [{...}] }, { method: 'post' })` will send as `FormData` with `myComplexData=[object Object]`.

**Reply by byudaniel** [reply]
The `submit` function only accepts simple objects with string values, see https://github.com/remix-run/remix/blob/5e16c8255bd0771aaf62b69c1a7ba6a4e773e4df/packages/remix-react/components.tsx#L999.

**Reply by meglio** [reply]
It got updated since then. It can now take a whole lot of different types as an input, including `FormData`:

![image](https://user-images.githubusercontent.com/282177/205428970-be9b245d-4e35-4ad9-9c88-5f29032b44f1.png)

**Comment by kiliman** [comment]
I've added a patch to add native JSON support to `useFetcher()`. I think this will help transition existing apps to Remix where there is heavy use of JSON APIs. The DX/UX of `useFetcher` state handling makes this superior to using plain `fetch()`.

```ts
export const action: ActionFunction = async ({ request }) => {
  const data = await request.json()
  return json({ result: { success: true, message: 'ok' }, posted: data })
}

export default function () {
  const fetcher = useFetcher()
  const data = fetcher.data
  const handleClick = () => {
    fetcher.submit(null, {
      action: '?index',
      method: 'post',
      json: { hello: 'world' },
    })
  }
  return (
    <div>
      <h1>useFetcher with JSON</h1>
      <pre>{JSON.stringify(data, null, 2)}</pre>
      <button onClick={handleClick}>Fetch</button>
    </div>
  )
}
```

https://github.com/kiliman/remix-json-fetcher

**Reply by jacob-ebey** [reply]
What about the API:

```js
fetcher.submit({ json: "values", go: "here" }, {
  action: '?index',
  method: 'post',
  json: true,
})
```

It feels strange to pass null where the "body" currently goes and have a config key for the JSON (body) to me.

**Reply by kiliman** [reply]
I was just trying to be minimal in my patch. The function was already doing a lot with the initial parameter. But I'm not tied to the API. So whatever works best is fine. We just need a way to indicate this is a JSON request. We could even overload `encType` to support `application/json`.

My original method was to do something like this, to be more explicit and match `fetch()`, but added too much boilerplate:
```ts
fetcher.submit(null, {
  action: '?index',
  method: 'post',
  body: JSON.stringify({ hello: 'world' }),
  headers: { 'content-type': 'application/json' }
})
```

**Reply by gabimor** [reply]
+1 for sending the json object as the first parameter

**Reply by lili21** [reply]
it doesn't work for the 1.11.0+ version.

**Comment by kiliman** [comment]
I submitted a PR to add this feature to `fetcher.submit()`

https://github.com/remix-run/remix/pull/1990

**Comment by ryanflorence** [comment]
Since the beginning I've been considering different enctypes for `<Form>`, and all of its cousins: `submit()`, `fetcher.Form`, `fetcher.submit()`.

So if I were to propose anything around this, the API would just be encType:

## Enctype all the things

```jsx
// Form API, any inputs w/ the same name become an array
<Form method="post" encType="application/json">
  <input name="email" />
  <input name="password" />
  <input type="checkbox" name="brand" value="nike" />
  <input type="checkbox" name="brand" value="adidas" />
</Form>

// useSubmit API, just plain JSON
submit(
  { email, name, brand: ["nike", "adidas"] },
  { method: "post", encType: "application/json" }
)
```

And of course, `fetcher.Form` and `fetcher.submit` would be the same.

My hesitation here is progressive enhancement, particularly early clicks. We can't just ignore it because w/o JavaScript the browser will ignore the invalid form enctype and send it as `"x-www-form-urlencoded"`. We also can't easily prevent those clicks without getting really weird.

## Only work with `submit` not `form`

If we only support `"application/json"` for `submit` and `fetcher.submit` then we don't have to worry about the early click problem.

This is probably the way to go.

```jsx
export function action({ request }) {
  if (request.headers.get("Content-Type") !== "application/json") {
    throw new Response("Unsupported Media Type", { status: 415 });
  }

  let json = await request.json();
  // carry on ...
}

export default function SomeRoute() {
  return (
    <button
      onClick={() => {
        submit(
          { email, name, brand: ["nike", "adidas"] },
          { method: "post", encType: "application/json" },
        );
      }}
    />
  );
}
```

## Back to lots of client state

> I've got a very complex form with nested arrays

One huge benefit of using HTML forms and `FormData` is that you can skip a ton of React state and let the DOM just deal with it. If you're getting values from the user you're now in charge of turning everything into controlled components and managing a bunch of React state to create that JSON object.

Is it worth it? I'm not convinced it is.

This usually comes up when migrating to Remix where you already had a ton of React state to build the object and did your own `fetch`. In those cases it's easy enough to dump all the JSON into a field and parse it on the server back to JSON:

```jsx
<input type="hidden" name="json" value={JSON.stringify(values)} />

// in the action
let json = JSON.parse(formData.get("json"))
```

If you take the time to refactor to form fields, I think you'll find you get delete a bunch of code and don't have to manage nearly as much state.

Also, don't forget that [formData already supports "arrays"](https://remix.run/docs/en/v1/pages/faq#how-can-i-have-structured-data-in-a-form) by naming the fields the same things and using `formData.getAll()`.

Anyway, I'm not opposed to this, but I personally need to see some examples of real world use cases that contrast both approaches. My gut tells me most folks will see the normal forms version and prefer it over managing all the form state manually to build up a JSON object.

**Reply by kiliman** [reply]
Me personally, I don't need to use it, but then again, most of my Remix projects are either new or complete rewrites. 

But there are many users coming to Remix and asking "How do I submit JSON?" because they want to migrate an existing project to Remix. For the most part, they aren't used to using standard forms and they have existing APIs that expect a JSON payload.

I just think that adding a simple flag to enable JSON will be a big win and reduce friction for those that want to try out Remix but don't want to jump all in.

**Reply by ryanflorence** [reply]
I think this proposal is a permanent middle ground with only half the benefits.

When migrating, you want to change as little code as possible. Switching a JSON fetch to a `useFetcher` w/ json content-type would likely break multiple parts of the app you're migrating. Instead, keep all of your form state management, global store, and manual JSON `fetch` (with the likely revalidation through the global store) "as is" and just move the server code to a remix Resource Route.

Once it runs in Remix, then switch over one form at a time to HTML/Remix APIs and get all the Remix network management benefits and delete a bunch of code.

I don't see how `useFetcher` w/ a JSON content type helps with migration.

**Reply by kiliman** [reply]
Ok. I guess I was arguing for a feature I wouldn't use anyway. I just know that based on Discord questions, this should probably be a FAQ somewhere.

**Reply by nicksrandall** [reply]
Thanks for explaining the trade-offs. The “real world use case” I have for this feature is I have a drag-and-drop interface where a user can sort a nested tree structure of data. On “dragend” I need to send the new tree structure to the the db to be saved. A tree structure like this can not (easily) be modeled as FormData. Only supporting JSON in submit method and not Form component would be enough for my needs. 

As a workaround, I am double encoding the data. I’m encoding as a JSON strong and then encoding again as FormData.

**Reply by dan-cooke** [reply]
>One huge benefit of using HTML forms and FormData is that you can skip a ton of React state and let the DOM just deal with it. If you're getting values from the user you're now in charge of turning everything into controlled components and managing a bunch of React state to create that JSON object.

To follow on from @nicksrandall 's point.

Another use case for when simple HTML form data is not a great fit is for "Wizards" . In this flow we typically _have_ to store users inputs in some object at the top level of the wizard. 

When it comes to finally submitting the full form - we tend to have nested objects here, ie. one object for each form wizard page. So FormData just doesn't work well for this.

I would love to see an example in the docs on how to implement a wizard with HTML forms. 

It would maybe be possible to have an action on each form step `/add/1` `add/2` etc. each of these actions stores the temporary form data server side, or encodes it in session storage or something. But it feels like mental gymnastics to me - It feels natural to be able to send JSON, although I do get the reasoning for preferring FormData

**Reply by jamesarosen** [reply]
> But there are many users coming to Remix and asking "How do I submit JSON?" because they want to migrate an existing project to Remix

This is not our case or our reasoning. Ours is a new app. We want JSON so we can pass arrays (already supported in Remix), objects (supported in Rails via nested `[]` naming), booleans, and numbers without having to manually parse them in the action.

There are ways to support `await response.json()` even with `x-www-form-urlencoded` data. For example, if the submitted form had
```
<input name="is_green" type="checkbox" />
<input name="__types[is_green]" type="hidden" value="boolean" />
```
then the action could use `__types` to do type coercion.

Remix could add the hidden inputs automatically via a custom `<TypedInput>` component and could wrap the `request` in such a way that `.json()` uses `__types` and nesting information to convert form-encoding to JSON.

**Later**

We can achieve most of the type-coercion we want with something like [Joi](https://www.npmjs.com/package/joi).

**Reply by meglio** [reply]
@ryanflorence Here is my use-case. In my project, I have a "price table editor", which, once upon clicking on the Save button, gives me a json object with quite a nesting. That would be ugly if expressed as form fields.

**Reply by cliffordfajardo** [reply]
In your form you could submit JSON like this in a standard & form compliant way & avoid needing to create multiple form fiends, for example:

```tsx
export async function action({ request, params }: ActionArgs) {
   const formData = await request.formData();
   const intent = formData.get("intent");
   switch (intent) {
      case "__update-user": {
          const updatePayload = JSON.parse(formData.get("body") as string);
           // do something with the data   
      }
     case "...": {...}
   }
       
}

// Component Code

fetcher.submit(
      {
        intent: "__update-user",
        body: JSON.stringify({ .....data....}),
      },
      { method: "post" }
    );
  };
```

that's how we're doing it in our app, but if there is an easier API that would facilitate the this whole process, that would be nice

Ryan's example is also how I would do it in other places:
```tsx
<input type="hidden" name="json" value={JSON.stringify(values)} />

// in the action
let json = JSON.parse(formData.get("json"))
```

**Reply by adanielyan** [reply]
I have yet another use case where form may not be easy to use. In my app users drag and resize different objects on the page (images, text, shapes, etc). There are no forms. JSON seems to be the most natural data format for my case.

**Reply by cliffordfajardo** [reply]
@adanielyan - in our app we don't use Remix's `<Form>` component at all because we use the `<Form>` component of the component library, so we use `submit` function a lot

Wouldn't useSubmit or usFetcher's `submit` function solve the JSON problem?
```tsx
const function handleDragEnd = () => {
   fetcher.submit(
      { intent: "__update-user", body: JSON.stringify({ .....someData....}),},
      { method: "post" }
   )
}

```

**Reply by adanielyan** [reply]
It would, however it's rather a workaround than a solution. My point was that not all use cases are a good fit for using forms and that there are cases where supporting json natively would make more sense.

**Reply by meglio** [reply]
Exactly, the solutions proposed are workarounds - those add frictions to the development process and make things more complicated instead of trying to simplify them for the developer. Yes maybe for 50%+ cases, forms are good. But you can't simply ignore a wide range of cases where you control a complex data struct and you want to send it to the server as is, without "native browser form serialization" mechanism taking place. You can't deny that more and more logic is put into the frontent, and it automatically means that you sometimes have more data and constructe it in a more sophysticated way to leave it to the "default" form behaviour. P.S. Also, I don't care about the "no-js-available" applications - that's exactly the small percentage of cases that should be easily sacrificed and not taken into account while considering a feature like this. At least in my own environment of developers, I have seen zero no-js applications in 18 years.

**Reply by dan-cooke** [reply]
> You can't deny that more and more logic is put into the frontent

this is the exact problem remix is trying to solve, and enforcing form data is a pretty good way Of achieving this in my opinion 

but I suppose it can’t hurt to support it for the cases were formdata is not a great fit

**Reply by adanielyan** [reply]
Enforcement never works...

**Reply by dan-cooke** [reply]
> Enforcement never works...

Wrong choice of words. Rather “recommending” form data is a great way of preventing framework users from shipping too much code.

obviously it cannot be enforced, users will always find ways around it

**Reply by adanielyan** [reply]
And I'm sure that's a valid recommendation for many (maybe most) use cases. But having an alternative for other use cases, while still recommending the form-data way, would be much more helpful. That's exactly the difference between enforcement and recommendation. You give people a recommendation and an option not to follow it if they don't want to.

**Reply by dan-cooke** [reply]
hence why i said 
> but I suppose it can’t hurt to support it for the cases were formdata is not a great fit

In my OP

**Reply by kiliman** [reply]
Since this is something that can be easily accomplished in userland, I don't think they'll modify the API to support native JSON.

Just create your own hook, `useJsonFetcher`.

**Reply by lili21** [reply]
> @adanielyan - in our app we don't use Remix's `<Form>` component at all because we use the `<Form>` component of the component library, so we use `submit` function a lot
> 
> Wouldn't useSubmit or usFetcher's `submit` function solve the JSON problem?
> 
> ```tsx
> const function handleDragEnd = () => {
>    fetcher.submit(
>       { intent: "__update-user", body: JSON.stringify({ .....someData....}),},
>       { method: "post" }
>    )
> }
> ```

It's still formData at network level. some in-app webview doesn't support sending formData

**Comment by dan-cooke** [comment]
### Is this a good idea?
When I first started using remix - this feature was something that I kept asking "Why didn't they implement this?" - but the more I used remix, I soon realised. This is by design.

Now I'm on the opposite side of the fence - **I don't think this should be supported at all.** 

### Why FormData is all we need
Its often said that a well designed language, framework has an established way of doing things - ideally a single way to do anything.

Well for handling user input remix has this sorted (or shall I say HTML had this sorted years ago, but we all forgot) - `FormData` and HTML Forms

Coming from a modern React application, our natural instinct is to turn our noses up at the use of `<form>` and `FormData`, its like the Rust compiler, at first you hate it - but you soon realise its just trying to help you.

If Remix adds first class support for Form JSON data type, it would encourage bad practices, in the use of controlled components and complex react state. 

### Why supporting JSON `encType` would lead to bad habits
It would allow developers coming from traditional react applications to merely use `useSubmit` or `useFetcher.submit` - pass it their massive state object with complex nested arrays, set `encType: 'application/json'` and wipe their hands with it. It will just work.

FormData is not so lenient. Nested arrays / objects are a no no with form data (or at least they aren't easy to work with). So you are forced to rethink your structure. 

### A recent example of mine
I was building a Step (Wizard) form. I built it with the typical React way, a huge state object tracking changes the user made at the top level. This object had multiple levels of nesting, and nested arrays. In order to even get this object sent to my `action` function, I had to really do some mental gymnastics - I had to use `fetch` and post to the route. It felt dirty, it felt hacky, it felt like I wasn't supposed to do that. It got me thinking - maybe I should do this differently? 

So I decided to completely isolate each page in the wizard, use HTML forms entirely, and store the progress of the wizard in session storage.

After my refactor I looked at the code diff, I had cut down the amount of code by almost half -not only that, but I am now shipping less to my users, I was able to remove `recoil` and `react-hook-form`. _And_ I get the added benefit of working without Javascript. 

Now I look back, and I am just in awe at the remix design philosphy - it made it difficult for me to do it "wrong". This is how frameworks and languages should be designed, the benefits should be abstracted away from you - you should get them for free, without thinking. Just by following the most basic standards.

Edit: so I think I have changed my mind minorly - I do think JSON should be supported now after reading various use cases on this thread. But I would be in favour of putting it low down in the docs and guiding people towards FormData for 90% of use cases

**Reply by gamliela** [reply]
> and store the progress of the wizard in session storage.

Can you elaborte on this point? what is actually stored in the session storage - the whole data from all pages, or just a reference to the current step?

**Reply by dan-cooke** [reply]
@IHIutch whatever you want, it’s up to you. I typically store the individual step data in separate keys

**Reply by gamliela** [reply]
okay, and then the last step in the wizard submits data from all steps together? How this is done without wrapping everything in a JSON?

I'm asking because I tried the approach of session storage too. It felt wrong to me because the form in each step has to be initialised with `useEffect` - because session storage cannot be accessed in the server loader.

**Reply by dan-cooke** [reply]
@gamliela im not referring to window.sessionStorage 

Check out remixes API for session storage which can be accessed on the server , using cookies or memory etc 

On mobile atm apologies

**Reply by gamliela** [reply]
ok, I used cookie storage for my wizard too. Problem is that storage is limited to 4k - might be too small if wizard involves textareas etc. Also cookie is shared with other routes, so space needs to be managed globally.

**Reply by dan-cooke** [reply]
In that case I’d recommend just using a database to store your state

**Comment by IHIutch** [comment]
Can someone ELI5 FormData for me in this case? I read this:
>It would allow developers coming from traditional react applications to merely use useSubmit or useFetcher.submit - pass it their massive state object with complex nested arrays, set encType: 'application/json' and wipe their hands with it. It will just work.
> FormData is not so lenient. Nested arrays / objects are a no no with form data (or at least they aren't easy to work with). So you are forced to rethink your structure.

And think, if it "will just work", why not support it? So that leads me to wonder: In what ways is FormData so superior that working with JSON becomes an anti-pattern?

Would love to learn more, thanks.

**Reply by dan-cooke** [reply]
Its not that its "superior" per say, for me the main benefit is that if it forces you to re-think your react code a little bit, and you end up with added benefit of supporting non-js users.

Take this example:

You have a multi-select screen with a tonne of cards the user can select via a checkbox.

In a "traditional" react app you would send this data to the server with JSON, your code would probably look something like this:

```ts
const [selectedIds, setSelectedIds] = useState();

<input id='1' type='checkbox' name="thing" checked={selectedIds.includes(1)} onChange={() => setSelectedIds(p => [...p, 1])} />
// and more 
```

With HTML forms and FormData you can't easily send an array over the wire, as you would have to append the data with something like

```.ts
const formData = new FormData();

selectedIds.forEach(id => formData.append('id', id))
```

Which just feels weird and you question it.

But instead you can use the HTML forms alongisde the FormData and remove all the state code, in remix you can just do 

```.ts
<form>
<input name='thing' type='checkbox' />
<input name='thing' type='checkbox' />
</form>

```
And when you submit this your action will receive the array `thing[]`

**Reply by IHIutch** [reply]
That's a helpful explanation, l appreciate you taking the time. If I'm understanding correctly, it seems that the recommendation toward using FormData is because it's a native web experience (via forms) where sending JSON is not. So in that sense JSON becomes the anti-pattern because it requires additional overhead to support?

(Putting my PHP hat on, this seems to make sense to me)

**Reply by IHIutch** [reply]
Following up here, the Remix docs pretty much address this in the FAQ: https://remix.run/docs/en/v1/pages/faq#how-can-i-have-structured-data-in-a-form

So while it's not technically JSON, you can still create a nested data structure and handle the parsing with a library like [query-string](https://github.com/sindresorhus/query-string). Given what's been mentioned before, that probably feels like an appropriate compromise while maintaining the integrity of "the platform"

**Reply by iway1** [reply]
Not going to argue about how absolutely wacky it is to not support JSON, just going to ask if could you at least change the docs to make it clear only form data is supported. The example given just shows passing an object and doesn't point out that actually you have to pass form data: https://remix.run/docs/en/v1/api/remix#usefetcher

Scroll down to fetcher.submit

**Reply by IHIutch** [reply]
You can submit the JSON as the example suggests, but it needs to be accessed via FormData in your Action.

Though I agree, this section of the docs aren't super clear to newcomers. Perhaps showing the `Action` receiving the data could help clarify.

**Reply by iway1** [reply]
Yeah, that would help a lot. Making Remix documentation more accessible would help with its adoption and save future devs some time (would've definitely saved me a bit of time).

I was getting a type error trying to pass an object as well which made me think I had to pass FormData to `submit`. Is that a bug with the typing? or does submit only allow string values ? My object was something like:

```
const o = {
    value: "Hi",
    numberValue: 5,
    boolValue: false
}
fetcher.submit(o, {method: 'post', action: '/something'}); // type error
```
I could do `//@ts-ignore` and it would submit, but feels bad man. If submit only likes string values that would be another great thing to include in the docs.

**Comment by lili21** [comment]
I get the point. Personally, I like FormData, I like the progressive enhancement philosophy. 
But in reality, FormData is not always an option. I'm building a project with remix right now, and the project needs to run inside App's Webview. and it doesn't support FormData for some security reason. I have to change it to JSON.

**Comment by barbinbrad** [comment]
As far as I understand this problem (forgive me for the ignorance) -- there is currently strong typing of data from loader to client, but from client to action the typing is not as strong. This is because `formData` can be a string or file.

The thing that brought me to this discussion was that I want to send a number from the client to the action without having to do `parseInt` inside the action. 

A nice workaround seems to be:

### `Client Side`
```tsx
const handleTeamSelect = (team: Identity) => {
    const data = new FormData();
    data.append("json", JSON.stringify(team));
    submit(data, { method: "post" });
 };
```

### `Action Side`
```tsx
export const action: ActionFunction = async ({ request }) => {
  const data = await request.formData();
  const identity = JSON.parse(data.get("json") as string);

  const identitySession = await getIdentitySession(request);
  identitySession.setIdentity(identity);

  return redirect("/home/overview", {
    headers: { "Set-Cookie": await identitySession.commit() },
  });
};
```

But even with this workaround, `identity` isn't strongly typed in the action.

**Reply by dan-cooke** [reply]
I tend to use a validation library like zod to validate form data. Parsing form data with a validator like this results in strong typing of your data

**Reply by barbinbrad** [reply]
Ahh, thanks so much! I see this sample repo, after grepping for zod: https://github.com/remix-run/remix/blob/cde451c5c1cc6dcaff74b8e6b4b7bf2e978ed2ec/examples/io-ts-formdata-decoding/README.md

**Reply by dan-cooke** [reply]
Awesome, I came across that example with io-ts before but haven’t tried it out.

I personally use a combination of `remix-validated-form` and `zod-form-data`.

But I like the minimal boilerplate in `io-ts`

**Reply by barbinbrad** [reply]
I've been using this suggestion for ~six months now. I can't recommend it enough! Thanks again @dan-cooke

**Comment by Klavionik** [comment]
Just wanted to add my 5 cents to the discussion. I understand the dev's point about nested data structures. However, what bothers me right now - how do we send a `null` to a server using FormData, if `null` is going to be converted to an empty string? 

Imagine we have such a case, when there's some REST API, and we need to update the entity's state with a PATCH request, sending something like that:

```json5
// PATCH /entity/1
{"expirationDate": null}
```

Why would we do that? For example if we have the `expirationDate` field of the entity set to some date and we want to "reset" it, effectively making the entity non-expirable. Our API doesn't accept empty strings, it considers them badly formatted dates. Only ISO 8601 strings and `null` are valid.

I've seen this pattern many times and as far as I understand, Remix (and React Router v6, which I'm trying to use) Actions don't support such cases. 

I'm sorry, if I'm missing something, but I wanted to point out this particular pain (I didn't find any earlier mention of this).

**Reply by allansson** [reply]
I think that if you don't set the name of the input or set it to empty, then the field will not be sent during submit. `FormData.get` will return null if a field does not exist.

```tsx
// Client
<input name={expirationDate === null ? '' : 'expirationDate'} value={expirationDate} />

// Server
const expirationDate = formData.get("expirationDate");
```

It is verbose though. You could always wrap the logic in a component to make it more readable.

**Reply by Klavionik** [reply]
Thank you for your suggestion, but it doesn't help in the described case. One need to specifically send a `null` value. Not sending the field at all just won't change the state.

**Comment by meglio** [comment]
I ended up using fetch() directly and ditching the beauty of the built-in `useFetcher()`. Unfortunately, but I'm forced to not use this Remix feature because of its design and restrictions.

**Reply by kiliman** [reply]
There's a new proposal for supporting different `encType` for serializing data. This includes support for `application/json`

_Better actions with opt-in serialization of payload data_
https://github.com/remix-run/react-router/discussions/10324

**Comment by vaynevayne** [comment]
1. No one will use raw input to write pages,
2. Most component libraries support value onChange controlled mode,
3. Generally, react-hooks such as useForm are used to manage the state of the table, so the controlled mode is the simplest.
4. As far as I know, people who write frameworks rarely touch composite business, so please keep the normal state in the design library, and don’t add too many personal ideas
5. for  formData i There's a reason why people don't use things that aren't popular, it is not family for devtool, it is too  terrible

**Comment by ZeldOcarina** [comment]
All right I guess I'll resort to good old Axios and setting all the loading states manually! 😅 
I still think it's weird to force users into that, I understand the principle, but json is also a native web feature, not being willing to support it is just weird, I use plain forms in 90% of my apps but when I have nested forms to create collections of very complex data it's just not ergonomic to do that..

**Reply by sergiodxa** [reply]
Support for sending JSON when using fetcher.submit or useSubmit has been implemented months ago in the [version v1.18.0 of Remix](https://github.com/remix-run/remix/releases/tag/remix%401.18.0)

**Reply by ZeldOcarina** [reply]
Oh nice to know, I'll get to it right away! Thank you!!! 😍🥰❤️

**Reply by ttsirkia** [reply]
So is that something which should be mentioned here https://remix.run/docs/en/main/hooks/use-fetcher#fetchersubmitformdata-options ?

**Reply by ZeldOcarina** [reply]
@ttsirkia tbh it's kind of there already, I went on to propose a change and it says it accepts the same options of useSubmit and if you check the useSubmit options you can see it there:

> **encType:** The encoding type to use for the form submission: application/x-www-form-urlencoded, multipart/form-data, application/json, or text/plain. Default is application/x-www-form-urlencoded.

**Reply by ttsirkia** [reply]
Yes, kind of. You just need to omit the information on that page and check the other page.

**Reply by ZeldOcarina** [reply]
Wanna propose a change? I would add a paragraph directly in the methods section specifying that.

**Reply by ttsirkia** [reply]
I can make a PR. The best place would be here?
![kuva](https://github.com/remix-run/remix/assets/2797729/6556c862-448a-4fd1-b5e2-d55dec360478)

Something like this:
Object can also be any JSON data if you pass `encType: "application/json"` option. See more details in `useSubmit` documentation.

**Reply by ZeldOcarina** [reply]
Oh yeah that would make sense!!

**Reply by ttsirkia** [reply]
PR #8787

**Comment by lopugit** [comment]
~~I have to agree it's a bit annoying that remix action's don't support `application/json`~~

~~It's a pretty fundamental feature of advanced web apps and API's..~~

**Reply by lopugit** [reply]
Oh goodness it's been implemented 😳 

Link: [version v1.18.0 of Remix](https://github.com/remix-run/remix/releases/tag/remix%401.18.0) 

IGNORE ME
